Feature Proposal Make Static Refactoring

CodeLtDave opened the pull request:
Fixes #590

Introducing a new Refactoring called "Make Static", a tool that transforms instance methods into static ones in the context of Eclipse's Java Development Tools. This involves adjusting method invocations appropriately to maintain logical correctness. The refactoring supports both instance and static contexts, passing 'this' for invocations within instance methods and the relevant instance variable for invocations in other classes or static methods. The refactoring utility aids in improving code modularity and reuse.

## What it does
The "Make Static Refactoring" transforms instance methods into static methods while ensuring the code remains logically correct. The refactoring process involves two key steps:

Modifying the method declaration: The instance method is changed into a static one by adjusting its declaration. If needed an additional parameter (an instance of the enclosing class) is added to the method signature to replace the usage of 'this' keyword in the method body.

Adjusting method invocations: All occurrences of the method invocations are modified to suit the new method signature. If the method invocation occurs in a static context, the relevant instance variable is passed as an argument. If the invocation is in an instance context, 'this' keyword is passed as an argument.

This refactoring utility is designed to improve code modularity and reuse by allowing methods to be invoked without the need for class instantiation.

## How to test
Utilize the provided MakeStaticRefactoringTests.java file, which includes various test cases for different scenarios. Run this file as a JUnit test to validate functionality.

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

jukzi commented:
That is much code to review. As i understand that feature should not break anything. Please prove that by applying it on the jdt.core repository and submitting that dogfooding PR.

jjohnstn commented on org.eclipse.jdt.ui/ui/org/eclipse/jdt/ui/actions/IJavaEditorActionDefinitionIds.java:
```diff
@@ -394,6 +394,14 @@ public interface IJavaEditorActionDefinitionIds extends ITextEditorActionDefinit
+	 * Action definition ID of the refactor -> make static action
+	 * (value <code>"org.eclipse.jdt.ui.edit.text.java.make.static"</code>).
+	 *
+	 * @since 3.31
```
This public api change will require bumping jdt.ui to 3.32.0 and this since tag will need changing to 3.32.

jjohnstn commented on org.eclipse.jdt.ui/ui/org/eclipse/jdt/ui/actions/JdtActionConstants.java:
```diff
@@ -392,6 +392,14 @@ public class JdtActionConstants {
+	 * Refactor menu: name of standard make static global action
+	 * (value <code>"org.eclipse.jdt.ui.actions.makeStatic"</code>).
+	 *
+	 * @since 3.31
```
similar bump needed to 3.32

jjohnstn commented on org.eclipse.jdt.ui.tests.refactoring/test cases/org/eclipse/jdt/ui/tests/refactoring/MakeStaticRefactoringTests.java:
```diff
@@ -0,0 +1,578 @@
+import org.eclipse.jdt.ui.tests.refactoring.infra.TextRangeUtil;
+import org.eclipse.jdt.ui.tests.refactoring.rules.RefactoringTestSetup;
+
+public class MakeStaticRefactoringTests extends GenericRefactoringTest {
```
This class needs to be added AllRefactoringTests.java

CodeLtDave replied:
"Thank you for pointingthat out. I've addressed this in 3eb9df5dea85910a1bdf5178386d8310b37e4140

jjohnstn commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/makestatic/ChangeCalculator.java:
```diff
@@ -0,0 +1,548 @@
+				ParameterizedType parameterizedType= (ParameterizedType) type;
+				List<Type> typeParamsOfMethodParam= parameterizedType.typeArguments();
+				for (Type typeParamOfMethodParam : typeParamsOfMethodParam) {
+					methodParameterTypes.add(typeParamOfMethodParam.resolveBinding().getName());
```
resolveBinding() can return null.  You need to check this here and other places to ensure you don't cause an NPE.

jjohnstn reviewed (changes requested):
Overall, looks good.  There are a few comments I have in the code.  One general one is to check the result of resolveBinding() for null before using it.

HeikoKlare commented:
> That is much code to review. As i understand that feature should not break anything. Please prove that by applying it on the jdt.core repository and submitting that dogfooding PR.

A dogfooding PR is a good idea. Should that only be for demonstration or with some real value, i.e., to actually be merged? In the latter case, I would propose to take a look at test classes, because they often contain utility methods defined as instance methods although they do not access the actual test instance. One example could be: `org.eclipse.jdt.core.tests.compiler.regression.AbstractRegressionTest#getCompilerTestsPluginDirectoryPath()`

I already applied the refactoring to some places in jdt.core and while most cases were handled properly, I have two findings:
1. I had one place that could not be refacored but threw an error because of "overlapping text edits". This is the case for
`org.eclipse.jdt.internal.codeassist.CompletionEngine#checkCancel()`.
2. It seems like a parameter for the original `this` object is always added to the method even if the method does not contain any instance access. Is that true and intended? That can, for example, be seen when refactoring
`org.eclipse.jdt.core.tests.compiler.regression.AbstractRegressionTest#getCompilerTestsPluginDirectoryPath()`. The refactoring adds an `AbstractRegressionTest` as parameter although I see no necessity for an accepting an instance of the class (and no usages of that parameter after applying the refactoring).

One final remark to the state of this PR: The contribution is the result of an internship in which we already did several reviews of the code before it was proposed here. For the sake of transparency, the reviews are still available: [Review 1](https://github.com/CodeLtDave/eclipse.jdt.ui/pull/3), [Review 2](https://github.com/CodeLtDave/eclipse.jdt.ui/pull/4), [Review 3](https://github.com/CodeLtDave/eclipse.jdt.ui/pull/5)

jjohnstn commented:
@HeikoKlare Regarding dogfooding, no we don't need to merge changes.  The point was to prove that the code doesn't simply break in real-world code.  I would like to get this code into 2023-03 M1 which has a deadline of Thursday this week but I do not see any responses to my review comments and it does not look like they have been implemented (e.g. the version bump required now that this patch has to go into 4.31).  I am also fine with having a few corner problem scenarios as we have M2 and M3 to solve them and to expose more. That is why we try and get large features like this into M1.  New issues would need to be opened as soon as the code is merged.

jjohnstn commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/makestatic/FinalConditionsChecker.java:
```diff
@@ -0,0 +1,208 @@
+		int parameterAmount= methodDeclaration.parameters().size() + 1;
+		String methodName= methodDeclaration.getName().getIdentifier();
+		IMethodBinding methodBinding= methodDeclaration.resolveBinding();
+		ITypeBinding typeBinding= methodBinding.getDeclaringClass();
```
Resolved

jjohnstn commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/refactoring.properties:
```diff
@@ -1345,6 +1345,30 @@ IntroduceIndirectionRefactoring_call_warning_anonymous_cannot_qualify=The declar
+MakeStaticRefactoring_method_is_overridden_in_subtype=Selected method is overridden in subtype. 
+MakeStaticRefactoring_hiding_method_of_parent_type=Make Static is not available. Method of parent type would be hidden after refactoring.
+MakeStaticRefactoring_not_available_for_recursive_methods=Make Static is not available for recursive methods.
+MakeStaticRefactoring_not_available_for_parametrized_methods=Make Static is not available for parametrized methods.
```
Resolved

jjohnstn commented:
Adding this to M1 to ensure proper time to test the feature.

jjohnstn commented:
@CodeLtDave Would you like to prepare a N&N entry for this new feature or would you like me to do it?

CodeLtDave commented:
> @CodeLtDave Would you like to prepare a N&N entry for this new feature or would you like me to do it?

I really appreciate the time you put in this pr. 

This is my first big pr to eclipse, can you elaborate what you mean with N&N entry? I think it would be a useful learning to me to do it myself if it is not too complicated. 

I was addressing the issues you found, how can I implement these?

jjohnstn commented:
Hi @CodeLtDave The @author tag issue can wait for M2 and requires a test.  I added resolveBinding checking and fixed the typo.  The issues that @HeikoKlare found need issues opened for M2.

The N&N entry is added to  www.eclipse.org-eclipse/news/4.31/jdt.html  You can read the instructions.html found in the news directory and see the previous N&N entries for 4.30 as an example.

HeikoKlare commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/makestatic/ChangeCalculator.java:
```diff
@@ -0,0 +1,548 @@
+			newParameterTag.setTagName(TagElement.TAG_PARAM);
+			newParameterTag.fragments().add(fTargetMethodDeclarationAST.newSimpleName(fParameterName));
+			ListRewrite tagsRewrite= fTargetMethodDeclarationASTRewrite.getListRewrite(javadoc, Javadoc.TAGS_PROPERTY);
+			tagsRewrite.insertFirst(newParameterTag, null);
```
If I verified correctly, the parameter documentation will always be the first javadoc entry with the current implementation (even before any "general" documentation). I propose to reuse functionality from existing refactoring implementations (such as "Change Method Signature") for a reasonable placement of the parameter documentation in javadoc. I have documented the issue here: #1043

HeikoKlare commented:
Thank you very much for rewiewing and merging the contribution, @jjohnstn!

I have documented all findings in dedicated issues:
- #1043 for the javadoc tag issue
- #1044 for the unnecessary parameter addition
- #1045 for the "overlapping text edits" failure

We will work on these issues to achieve an improved state of the refactoring for M2.

HeikoKlare commented:
For the record: I have put some more effort into verifying the behavior of the "Make Static" refactoring. To this end, I have applied the refactoring to every method within the class `org.eclipse.jdt.core.tests.compiler.regression.AbstractRegressionTest` for which the preconditions are fulfilled (e.g., not being an overwritten or overwriting method).

The results are as follows:
* `Compilation`: Only a single compile error occurs. The reason is documented in #1083 and fixed in #1085. With that fix, no compile error occurs.
* `Behavior`: I have executed the regression test suite (`org.eclipse.jdt.core.tests.compiler.regression.TestAll`), in which many test classes extend the refactored `AbstractRegressionTest`. I have 70 test failures (out of more than 100k tests) in my local setup, but they are exactly the same for the code as-is and the refactored code, so no changes in behavior because of the refactoring.

jukzi commented:
> * I have executed the regression test suite

the good thing about dogfooding PR would be that all those checks come for free.

HeikoKlare commented:
> the good thing about dogfooding PR would be that all those checks come for free.

Thanks for refreshing the notice on a dogfooding PR, @jukzi! Sorry that I forgot about that one after the feature had been merged before I made that PR. I have opened the according PR for JDT Core: eclipse-jdt/eclipse.jdt.core#1878
Everything compiles and all tests (in particular the [regression tests](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-1878/1/testReport/org.eclipse.jdt.core.tests.compiler.regression/) affected by the refactoring) succeed.
